gh-115528: Update language reference for PEP 646

mrahtz opened the pull request:
<!-- gh-issue-number: gh-115528 -->
* Issue: gh-115528
<!-- /gh-issue-number -->


<!-- readthedocs-preview cpython-previews start -->
----
📚 Documentation preview 📚: https://cpython-previews--120294.org.readthedocs.build/

Changed sections:
* https://cpython-previews--120294.org.readthedocs.build/en/120294/reference/compound_stmts.html#function-definitions
* https://cpython-previews--120294.org.readthedocs.build/en/120294/reference/expressions.html#subscriptions
* https://cpython-previews--120294.org.readthedocs.build/en/120294/reference/expressions.html#expression-lists

<!-- readthedocs-preview cpython-previews end -->

mrahtz committed:
Update language reference for PEP 646

AlexWaygood commented on Doc/reference/compound_stmts.rst:
```diff
@@ -1336,6 +1338,10 @@ enables postponed evaluation.  Otherwise, they are evaluated when the function
 
+.. versionchanged:: 3.11
+   Parameters of the form "``*identifier``" may have an annotation
+   "``: *expression``". See :pep:`646` for details.
```
It's fine to link to PEP-646 from these docs, as PEP-646 explains the motivation and history behind this change in a complete way. But nowadays we prefer to have all relevant details in the CPython docs, and for PEPs to be seen as historical documents that are generally not meant to be updated after they're accepted. So I'd prefer us to avoid, if possible, the implication that there are any details in PEP-646 that aren't included in the CPython docs, because ideally there shouldn't be any.

AlexWaygood replied:
(I know lots of other bits of the docs say things like "See PEP XXX for details", but this is something we're trying to slowly move away from!)

JelleZijlstra replied:
I think a versionchanged note is a good place to link a PEP, because it provides background and context for the change. Perhaps we can simply:
```suggestion
   "``: *expression``". See :pep:`646`.
```
to remove the implication that the PEP has more information specifying the behavior.

JelleZijlstra commented on Doc/reference/expressions.rst:
```diff
@@ -1874,7 +1878,7 @@ Expression lists
 
 .. productionlist:: python-grammar
-   expression_list: `expression` ("," `expression`)* [","]
+   expression_list: `starred_expression` ("," `starred_expression`)* [","]
```
This is not strictly correct. `expression_list` according to the grammar is what appears after `yield`, but `def f(): yield *a` is a SyntaxError. (However, `yield a, *b` is legal.)

mrahtz committed:
Attempt 2

To recap: the objective is to make starred expressions valid in `subscription`,
which is used for generics: `Generic[...]`, `list[...]`, etc.

What _is_ gramatically valid in such contexts? Seemingly any of the following.
(At least, none of the following throw `SyntaxError` in a 3.12.3 REPL.)

    Generic[x]
    Generic[*x]
    Generic[*x, y]
    Generic[y, *x]
    Generic[x := 1]
    Generic[x := 1, y := 2]

So introducting

    flexible_expression: expression | assignment_expression | starred_item

end then switching `subscription` to use `flexible_expression` sorts that.

But then we need to field `yield` - for which any of the following are
apparently valid:

    yield x
    yield x,
    yield x, y
    yield *x,
    yield *x, *y

Introducing a separate `yield_list` is the simplest way I've been figure out to
do this - separating out the special case of `starred_item ,`.

mrahtz committed:
Cleanups

mrahtz committed:
Simplify
